Enable SVE in ISA-L erasure code for aarch64

docularxu opened the pull request:
This patch adds Arm (aarch64) SVE [1] variable-length vector assembly support
into ISA-L erasure code library. "Arm designed the Scalable Vector Extension
(SVE) as a next-generation SIMD extension to AArch64. SVE allows flexible
vector length implementations with a range of possible values in CPU
implementations. The vector length can vary from a minimum of 128 bits up to
a maximum of 2048 bits, at 128-bit increments. The SVE design guarantees
that the same application can run on different implementations that support
SVE, without the need to recompile the code. " [3]

Test method:
 - This patch was tested on Fujitsu's A64FX [2], and it passed all erasure
     code related test cases, including "make checks" , "make test", and
     "make perf".
 - To ensure code testing coverage, parameters in files (erasure_code/
     erasure_code_test.c , erasure_code_update_test.c and gf_vect_mad_test.c)
     are modified to cover all _vect versions of _mad_sve() / _dot_prod_sve()
     rutines.

Performance improvements over NEON:
In general, SVE benchmarks (bandwidth in MB/s) are 40% ~ 100% higher than NEON
when running _cold style (data uncached and pulled from memory) perfs. This
includes routines of dot_prod, mad, and mul.

Optimization points:
This patch was tuned for the best performance on A64FX. Tuning points being
touched in this patch include:
1) Data prefetch into L2 cache before loading. See _sve.S files.
2) Instruction sequence orchestration. Such as interleaving every two
     'ld1b/st1b' instructions with other instructions. See _sve.S files.
3) To improve dest vectors parallelism, in highlevel, running
     gf_4vect_dot_prod_sve twice is better than running gf_8vect_dot_prod_sve()
     once, and it's also better than running _7vect + _vect, _6vect + _2vect,
     and _5vect + _3vect. The similar idea is applied to improve 11 ~ 9 dest
     vectors dot product computing as well. The related change can be found
     in ec_encode_data_sve() of file:
     erasure_code/aarch64/ec_aarch64_highlevel_func.c

Notes:
1) About vector length: A64FX has a vector register length of 512bit. However,
     this patchset was written with variable length assembly so it work
     automatically on aarch64 machines with any types of SVE vector length,
     such as SVE-128, SVE-256, etc..
2) About optimization: Due to differences in microarchitecture and
     cache/memory design, to achieve optimum performance on SVE capable CPUs
     other than A64FX, it is considered necessary to do microarchitecture-level
     tunings on these CPUs.

[1] Introduction to SVE - Arm Developer.
      https://developer.arm.com/documentation/102476/latest/
[2] FUJITSU Processor A64FX.
      https://www.fujitsu.com/global/products/computing/servers/supercomputer/a64fx/
[3] Introducing SVE.
      https://developer.arm.com/documentation/102476/0001/Introducing-SVE

Signed-off-by: Guodong Xu <guodong.xu@linaro.org>

docularxu commented:
That's a false format report, isn't it? I had my signed-off-by in the commit message, which goes “Signed-off-by: Guodong Xu <guodong.xu@linaro.org>”.
![Screenshot from 2021-12-29 06-34-19](https://user-images.githubusercontent.com/7981735/147612068-4d084cde-0daa-4ee5-9b31-db779b751d2f.png)
 However, the CI / check_format still reports that there is a signed-off-by error.

Can somebody help me with more details?

gbtucker commented:
> That's a false format report, isn't it? I had my signed-off-by in the commit message, which goes “Signed-off-by: Guodong Xu [guodong.xu@linaro.org](mailto:guodong.xu@linaro.org)”. However, the CI / check_format still reports that there is a signed-off-by error.
> 
> Can somebody help me with more details?

I'm not sure.  From the log it seems to be checking out the wrong commit
```
Deleting the contents of '/home/runner/work/isa-l/isa-l'
Initializing the repository
Disabling automatic garbage collection
Setting up auth
Fetching the repository
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +c022350f173b3e5d6c5be559770b31297e6eb400:refs/remotes/pull/196/merge
 From https://github.com/intel/isa-l
   * [new ref]         c022350f173b3e5d6c5be559770b31297e6eb400 -> pull/196/merge
Determining the checkout info
Checking out the ref
/usr/bin/git log -1 --format='%H'
'c022350f173b3e5d6c5be559770b31297e6eb400'
```
That is not the commit ID of the latest PR however.  If I had to guess it might be because we have just recently added the commit for the github CI workflow and this doesn't appear in your git history yet.  Perhaps a rebase with upstream will fix.

docularxu commented:
> That is not the commit ID of the latest PR however. If I had to guess it might be because we have just recently added the commit for the github CI workflow and this doesn't appear in your git history yet. Perhaps a rebase with upstream will fix.

That reminded me. Now I did a rebase to the latest master. However, the check_format failure is still there.

You are right that the github CI is checking the wrong 'commit id'. From the log, it's checking the 'merge' patch, not the one I submitted.

docularxu commented:
Ping?

gbtucker commented:
I've fixed the issue with signoff checking with latest push (642ef36).

docularxu commented:
> I've fixed the issue with signoff checking with latest push ([642ef36](https://github.com/intel/isa-l/commit/642ef36874c59903f3bdef5964a2f8b9b50f3df0)).

Thanks. I rebased.

yuhaoth reviewed (approved):
I have checked c source code and assembly code. It looks good to me

gbtucker commented:
Integrated
